File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/MailResponseProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 邮件(返回) </summary>
public struct MailResponseProto : IProto
{
    public ushort ProtoCode { get { return 50002; } }

    /// <summary> 邮件数量 </summary>
    public int MailCount;
    /// <summary> 邮件项 </summary>
    public List<MailItem> MailList;

    /// <summary> 邮件项 </summary>
    public struct MailItem
    {
        /// <summary> 邮件Id </summary>
        public int Id;
        /// <summary> 邮件内容 </summary>
        public string Content;
        /// <summary> 是否已读 </summary>
        public bool IsRead;
    }

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(MailCount);
            for (int i = 0; i < MailCount; i++)
            {
                ms.WriteInt(MailList[i].Id);
                ms.WriteUTF8String(MailList[i].Content);
                ms.WriteBool(MailList[i].IsRead);
            }
            return ms.ToArray();
        }
    }

    public static MailResponseProto GetProto(byte[] buffer)
    {
        MailResponseProto proto = new MailResponseProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.MailCount = ms.ReadInt();
            proto.MailList = new List<MailItem>();
            for (int i = 0; i < proto.MailCount; i++)
            {
                MailItem _Mail = new MailItem();
                _Mail.Id = ms.ReadInt();
                _Mail.Content = ms.ReadUTF8String();
                _Mail.IsRead = ms.ReadBool();
                proto.MailList.Add(_Mail);
            }
        }
        return proto;
    }
}

File: GameServer/GameServerApp/GameServerApp/Proto/CreateLuaProto/MailResponseProto.lua
--邮件(返回)
MailResponseProto = { ProtoCode = 50002, MailCount = 0, MailTable = { } }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
MailResponseProto.__index = MailResponseProto;

function MailResponseProto.New()
    local self = { }; --初始化self
    setmetatable(self, MailResponseProto); --将self的元表设定为Class
    return self;
end


--定义邮件项
Mail = { Id = 0, Content = "", IsRead = false }
Mail.__index = Mail;
function Mail.New()
    local self = { };
    setmetatable(self, Mail);
    return self;
end


--发送协议
function MailResponseProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.MailCount);
    for i = 1, proto.MailCount, 1 do
        ms:WriteInt(MailList[i].Id);
        ms:WriteUTF8String(MailList[i].Content);
        ms:WriteBool(MailList[i].IsRead);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function MailResponseProto.GetProto(buffer)

    local proto = MailResponseProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.MailCount = ms:ReadInt();
	proto.MailTable = {};
    for i = 1, proto.MailCount, 1 do
        local _Mail = Mail.New();
        _Mail.Id = ms:ReadInt();
        _Mail.Content = ms:ReadUTF8String();
        _Mail.IsRead = ms:ReadBool();
        proto.MailTable[#proto.MailTable+1] = _Mail;
    end

    ms:Dispose();
    return proto;
end